File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Odd_Even_Linked_List/yujin.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def oddEvenList(self, head: Optional[ListNode]) -> Optional[ListNode]:
        odd = True
        odd_list = odd_head = None
        even_list = even_head = None
        cur = head
        while cur:
            if odd:
                if odd_list is None:
                    odd_list = odd_head = cur
                else:
                    odd_list.next = cur                    
                    odd_list = odd_list.next
                odd = False
            else:
                if even_list is None:
                    even_list = even_head = cur
                else:
                    even_list.next = cur
                    even_list = even_list.next
                odd = True
            cur = cur.next
        if odd_list is not None:
            odd_list.next = even_head
        if even_list is not None:
            even_list.next = None

        return head

File: VIF_git/GSDS-Qual/LeetCode/week1/Linked_List/Odd_Even_Linked_List/yujin.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* oddEvenList(ListNode* head) {
        ListNode* oddList = nullptr;
        ListNode* evenList = nullptr;
        ListNode* evenHead = nullptr;

        ListNode* cur = head;
        bool odd = true;
        while (cur) {
            if (odd) {
                if (oddList == nullptr) {
                    oddList = cur;
                }
                else {
                    oddList->next = cur;
                    oddList = oddList->next;
                }
                odd = false;
            }
            else {
                if (evenList == nullptr) {
                    evenList = evenHead = cur;
                }
                else {
                    evenList->next = cur;
                    evenList = evenList->next;
                }
                odd = true;
            }
            cur =cur->next;
        }
        if (oddList != nullptr) {
            oddList->next = evenHead;
        }
        if (evenList != nullptr) {
            evenList->next = nullptr;
        }
        return head;
    }
};
